cbmc_viewer/clauset.py

```diff
@@ -0,0 +1,263 @@
+# Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+"""CBMC solver query complexity metric.
+
+This module displays mapping between clauses and program code.
+It includes SSA expression corresponding to the line of code, respective
+clauses and the number of clauses, variables and literals for the LOC.
+Additionally, it tracks clauses in the UNSAT core and displays LOC that do
+not contribute any clauses to the core.
+Inputs to the module is a json file obtained by running CBMC with the
+--write-solver-stats-to parse option, dimacs formula from --dimacs CBMC
+option, the UNSAT core obtained by running kissat (to get proof of UNSAT)
+and then drat-trim (extracts the core), the project source and reporting
+directories.
+"""
+
+import logging
+import json
+
+import voluptuous
+import voluptuous.humanize
+
+from cbmc_viewer import filet
+from cbmc_viewer import parse
+from cbmc_viewer import srcloct
+from cbmc_viewer import templates
+from cbmc_viewer import util
+
+JSON_TAG = "viewer-clause"
+
+################################################################
+
+VALID_CLAUSE = voluptuous.schema_builder.Schema({
+    'GOTO': str, # string storing the goto instruction
+    'GOTO_ID': int, # goto instrucion ID as assigned by CBMC for the above instruction
+    'SAT_hardness': { # stores solver complexity stats for this instruction
+        'ClauseSet': [[int]],
+        'Clauses': int,
+        'Literals': int,
+        'Variables': int
+        },
+    'SSA_expr': str,
+    'Source': srcloct.VALID_SRCLOC
+    },required=True)
+
+VALID_CLAUSE_SUMMARY = voluptuous.schema_builder.Schema({
+    'summary': [VALID_CLAUSE], # list of instructions with the corresponding solver query stats
+    'core': {
+        'instr': [int], # list of goto ID of instructions that contribute to core
```

Comment:
The core is a list of clauses.  We map those clauses to instructions.  This core summary that is generated splits instructions into those contributing to the core that those that don't.  To find the clause corresponding to an instruction with id I in the core, scan the list of instructions in summary for an instruction with id I.